docs/signature.md

```diff
@@ -0,0 +1,93 @@
+# Image Signature Specification
+
+**Version 0.1**
+
+## Introduction
+
+This document defines a detached container image signature object and signing methods.
+
+## Signature Format
+
+```js
+{
+    "critical": {/* required fields */
+        "identity": {/* identity reference */},
+        "image": {/* signed object reference */ },
+        "type": "..."
+    },
+    "optional": {/* optional metadata fields */}
+    }
+}
+```
+
+### Fields
+
+There are two top-level fields, **critical** (required) and **optional** (optional).
+
+#### `critical`
+
+**identity** (string):
+
+```js
+{
+    "docker-reference": imageName
+}
+```
+
+`imageName` per [V2 API](https://docs.docker.com/registry/spec/api/#/overview) Required.
```

Comment:
On Fri, Nov 18, 2016 at 02:05:53PM -0800, Miloslav Trmač wrote:
> > You can avoid this by not including transport/location information in the asserted name.
>
> That doesn’t really work; we need a global namespace for looking up the trusted key (cryptographic signature verification should happen before anything even _touches_ the JSON data!)…

What does this have to do with JSON?  You should only be looking inside the name-assertion after you've authenticated the signature on it as “signed by someone you trust to assert the name you expect to see inside”.  Decoupling the transport from the asserted name shouldn't affect that.  And if you're comfortable with signature-discovery as layed out in #120, there's no reason you need to put the CAS- or ref-engine location in the *name* to use that scheme.  Just use internal logic like:

    signature_location(engine_location, requested_name) → URLs

> … and we need to be able to compare the name inside the signature with something the user has provided which unambiguously specifies intent.

Agreed.  I think “RHEL-8.10.5” unambiguously specifies intent.

> It is acceptable to ask the user to do an one-time setup to trust key X for example.com; it is completely unreasonable to ask the user to say `$tool pull --expected-name='Product By Example.COM 5.1.3' example.com/product5/standalone:1.3`.

I'd name that product “product5/standalone:1.3”.  Then the user runs:

  $ tool pull https://example.com product5/standalone:1.3

which says “hit the refs and CAS engines at https://example.com and fetch me ‘product5/standalone:1.3’”.  You could also specify default ref/CAS engines in a config file and run:

  $ tool pull product5/standalone:1.3                                                                                                                                                                                                          

which says “go to wherever you usually go for this sort of thing and fetch me ‘product5/standalone:1.3’”.